KFSPTS-20397 Use AWS Secrets to manage Concur credentials

cah292 opened the pull request:
There's a related PR for this in the "config" project. That other PR may need to be merged before the unit tests will pass, since it introduces some config properties that may need to be present for KFS to launch.

At the moment, I have not found any active SQL in "nonprod-sql" that updates or removes Concur web service credential entries, so I have not created a "nonprod-sql" PR yet. If you end up finding such SQL, please let me know.

This PR updates the ConcurAccessTokenService to use the AWS Secrets Manager for handling the Concur credentials. I've divided the credentials into two JSON objects: One for a "static" config containing properties that are not expected to change, and one for a "token" config containing the token-specific details that change frequently. Also, note that the service has been updated to change the access token on each token refresh, to correctly account for a Concur change noted in the user story.

In addition, this PR introduces several new unit-test-related classes, divided into two main areas. The simpler of the two is a "mock" AWS Secret Service, which is backed by a Java Map instead of actually invoking the Secrets Manager. This allows for easier micro-testing of the code's interaction with AWS secrets.

The more complex addition is some new test classes for improved JAX-RS-related unit testing. The classes rely on Jersey's core test framework, along with some custom code that allows for using an Apache HTTP server to handle requests to a localhost JAX-RS endpoint. I went with the Apache HTTP setup to keep the web processing more efficient (given the good performance we've gotten with similar tests in the past), and to keep the number of new test dependencies to a minimum. If needed, we could look into using one of Jersey's other provided JAX-RS container implementations in the future.

Please let me know if you have questions or concerns about this code, especially the new JAX-RS unit testing classes.

jhulslander reviewed (commented):
I'm still reviewing this PR, but I would suggest adding a unit test to verify POJO marshaling and unmarshaling from JSON that will be stored in the AWS secret

cah292 commented:
As requested, I've added some unit tests to verify the JSON marshal/unmarshal of the new Concur POJOs. I've also made a few other improvements to the tests, and added a new unit test for validating the correctness of a test utility class introduced by this PR.

This is ready for review again.

jhulslander commented on src/test/java/edu/cornell/kfs/sys/rest/util/ApacheHttpJerseyExampleTest.java:
```diff
@@ -0,0 +1,145 @@
+import edu.cornell.kfs.sys.rest.resource.TestJaxrsResource;
+
+@Execution(ExecutionMode.SAME_THREAD)
+public class ApacheHttpJerseyExampleTest {
```
Is this example needed?  It looks like ConcurAccessTokenServiceImplTest exercises ApacheHttpJerseyTestExtension

cah292 replied:
As long as we expect at least one other test (like the Concur one) to remain around to exercise the ApacheHttpJerseyTestExtension code, I'm fine with removing this example test. I'll go ahead and remove this for now.

jhulslander replied:
Lets bring this up in the stand up tomorrow.  If there is a user story for the next usage of AWS secrete manager, a not should be added to make a unit test similar to this.

jhulslander replied:
Where do we stand with this?  I thought we wanted to remove this example since we have working tests for Concur?

cah292 replied:
I agree that the Concur code is exercising enough of this functionality for now. I'll proceed with removing this particular test.

jhulslander commented on src/test/java/edu/cornell/kfs/sys/service/mock/MockAwsSecretServiceImplTest.java:
```diff
@@ -0,0 +1,154 @@
+    @BeforeEach
+    void setUp() throws Exception {
+        awsSecretServiceImpl = new MockAwsSecretServiceImpl();
+        awsSecretServiceImpl.setAwsRegion(MockAwsSecretServiceConstants.AWS_US_EAST_ONE_REGION);
```
As we move more functionality into AWS Secrets Manager, I expect MockAwsSecretServiceImpl will be used more often.  Do you think setting the region, namespaces and retry count could be moved into the constructor to make reuse easier?

cah292 replied:
Ok, I'll add more constructor options to make the mock secret service easier to set up.

jhulslander reviewed (changes requested):
I found a few items to take a look at.  The jaxrs test framework is pretty involved and looks good for the most part.  I think it would be good to share in a bookclub or something.
The one major item of the code review is adding some caching of AWS secrets to minimize calls out to AWS secrets manager.

cah292 commented on src/main/java/edu/cornell/kfs/concur/service/impl/ConcurAccessTokenServiceImpl.java:
```diff
@@ -48,25 +53,49 @@
+        try {
+            awsSecretService.updatePojo(ConcurAwsKeyNames.TOKEN_CONFIG, true, tokenConfig);
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException(e);
```
Ok, I'll add some logging here.

cah292 replied:
Yes, I'll add logging here. Sorry that I missed that change last time. (I thought I added it in, but realized I had edited the method above it that was also missing the logging. At least now both of those methods will have the logs added in.)

cah292 commented on src/main/java/edu/cornell/kfs/concur/aws/ConcurTokenConfig.java:
```diff
@@ -0,0 +1,31 @@
+public class ConcurTokenConfig {
+    private String access_token;
+    private String refresh_token;
+    private String access_token_expiration_date;
```
We were formerly using the expiration date as a string, since we were storing it in the Web Service Credentials table and that only allows for strings. However, now that your new AwsSecretService is available and can auto-parse DTOs accordingly, I agree that we could now store this value as a Date instead. I'll adjust the DTO and its related code accordingly.

cah292 commented on src/main/java/edu/cornell/kfs/concur/service/impl/ConcurAccessTokenServiceImpl.java:
```diff
@@ -48,25 +53,49 @@
     protected ConcurBatchUtilityService concurBatchUtilityService;
 
+    protected ConcurStaticConfig getStaticConfig() {
+        return getConcurPojoFromAws(ConcurAwsKeyNames.STATIC_CONFIG, ConcurStaticConfig.class);
```
I could add instances variables instead of static variables to cache the tokens, but it would be helpful to have some way to manually or periodically flush the cache, in case we have a need to manually change the token through the UI or the Secrets Manager. Perhaps I could add a service to the bus (or set up a cache with EHCache) that could clear out or re-retrieve the tokens via the KFS Concur screen, and that would at least take care of the KFS runtime instances. As for KFS Batch, we could just request a batch container restart if we need it to pick up a manual token change.

jhulslander replied:
That is a good point about refreshing on demand.  What if the cache value is "scoped" to a job run?  Meaning when the job run it pulls the secrets from AWS, then using the in memory values for the rest of the job run.  Then start fresh again.

cah292 replied:
It may not be practical to "scope" the tokens on a per-job-run basis, due to the complexities it would introduce for determining when and how the "scope" starts or ends.

After some further thought, there could be another approach that would work around the difficulties with the EHCache distributed caching. I could set up a local in-memory cache of AWS secrets, but they would calculate a special string from the secret value (such as a checksum or timestamp), and then store the special value in the database (such as through the Web Service Credential service or a brand new service). Later, when retrieving a secret value from the cache, it can compare the locally-stored timestamp/checksum against the DB-stored one, to determine whether the secret value needs to be re-retrieved from AWS.

Please let me know your thoughts about this, or if you think this is introducing an unnecessary amount of extra caching code.

jhulslander replied:
I think we are trending towards too complicated with that approach.  I was looking towards reducing the number of times we had to retrieve a secret from AWS, but if we over complicate the process, extra time spent supporting it would over shadow savings.

cah292 replied:
In that case, I'm not aware of an appropriate alternative, unless we want a per-instance cache that just expires its elements after a certain amount of time. If there's enough spare time left at the end of tomorrow's MOPS meeting, we could check with the rest of the team to get their thoughts on what (if any) AWS secrets caching would be best. (We might also be able to use that time to discuss your other concern about the example-jaxrs-server unit test.)

cah292 commented on src/main/java/edu/cornell/kfs/concur/service/impl/ConcurAccessTokenServiceImpl.java:
```diff
@@ -48,25 +53,49 @@
+
+    protected <T> T getConcurPojoFromAws(String awsKeyName, Class<T> pojoClass) {
+        try {
+            return awsSecretService.getPojoFromAwsSecret(awsKeyName, true, pojoClass);
```
I'm wondering if we should go with a hybrid approach for the shared-vs-instance usage of the new DTOs. The static-config DTO could be the shared one that is used by all KFS environments. However, if we set up the token-config DTO to be instance-specific, that would allow each non-PROD environment (and our local development environments) to independently set up and manage their own access tokens. This could simplify testing the Concur Event Notification process in non-PROD environments, and will also avoid the risk of accidentally having a non-PROD environment change the PROD token.

Please let me know your thoughts on whether a shared-static-config and instance-token-config setup makes sense.

jhulslander replied:
I like the idea of being able to test in dev and test independently from prod.  I think we would need to cooridinate with Tami to set up the accounts

cah292 commented on src/main/webapp/jsp/concur/manageAccessToken.jsp:
```diff
@@ -37,8 +37,8 @@
-					<div style="font-weight: bold">Reset existing token to empty string in the KFS database.</div>
-					<div>(Do this in non-prod databases to test revoke access token.)</div>
+					<div style="font-weight: bold">Reset existing token to empty string in the back-end storage.</div>
+					<div>(Do this in non-prod systems to test revoke access token.)</div>
```
See my ConcurAccessTokenServiceImpl notes on potentially keeping the token part of the config as instance-specific. If we took that approach, then we could safely allow the revoking of environment-specific tokens. However, I'll likely still need to make a few other adjustments to these buttons anyway, to address the new caching code that will be added.

cah292 commented:
I've committed some more changes to address a few of the items from the current feedback. There are still more areas that I need to address, including the potential caching part that our team will discuss at tomorrow's meeting.

In addition, when converting the expiration date handling to treat the value as a Java Date, I discovered an issue where the Jackson JSON processing seems to be defaulting to an unexpected time zone when handling dates. I've added some code that will configure the ObjectMapper to work around the problem, and have added a few unit test cases that can validate my fix for the issue.

This is not ready for final review yet, but it can still be reviewed for tomorrow's cache-setup meeting.

cah292 commented:
Also, I've rebased this feature branch to resolve a merge conflict. If you notice anything odd with the code as a result of that, please let me know.

cah292 commented:
I've committed some more changes to start implementing the AWS cache handling that was discussed at our team meeting back on Wednesday.

In its current state, the new caching uses functionality and coding that is similar to the financials GlobalVariables.doInNewGlobalVariables() method, but which is independent of that financials method. That way, we don't have to worry about carrying over the user session or message map to the cache-enabled code. This setup also allows for improved exception handling, so that the calling code is not required to always deal with the base Exception type.

Most of the new coding is complete, but I still need to perform some more local testing before the code can be merged. I may also need to make a few more changes to the unit test code or runtime code.

Please let me know if you think the caching strategy here fits with what we were discussing back on Wednesday, or if you have some suggestions for improving or replacing it.

jhulslander commented on src/main/java/edu/cornell/kfs/sys/util/CuJsonUtils.java:
```diff
@@ -0,0 +1,60 @@
+        }
+    }
+
+    public static String rebuildJsonStringWithPropertyOverride(
```
Is this function needed?  It looks like it is only used in testing.  I am not sure if we would need the functionality. 

The current functionality adds new elements to existing json.  The name implies to me that it will change the value of an existing element in the json.  If you think this function is needed for this PR, I think it should be renamed.

cah292 replied:
This method and the buildJsonStringFromEntries() method are meant for unit testing convenience instead of Production code, though the buildObjectMapperUsingDefaultTimeZone() method is used in Production code. I will create a separate CuJsonTestUtils class to distinguish between the two categories.

As for this method's functionality, it will overwrite the JSON property if it exists, otherwise it will add it to the object. I'll see if I can rename this to make the intent clearer.

jhulslander commented on src/test/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImplIntegrationTest.java:
```diff
@@ -11,13 +11,18 @@
+@AwsSecretServiceCacheExtension(awsSecretServiceField = "awsSecretServiceImpl")
 class AwsSecretServiceImplIntegrationTest {
     private static final String AWS_US_EAST_ONE_REGION = "us-east-1";
     private static final String KFS_INSTANCE_NAMESPACE = "kfs/local-dev/";
```
This PR contains a sys constants class with these same values.  Should these local variables be replaced with what is in the constants class?

cah292 replied:
Good catch, I'll clean up these variables and use the ones from the new constants class instead.

jhulslander reviewed (changes requested):
changes look good, and the cacheing seems reasonable .... I noted a few things to consider. Also, did  we want to remove the code related apache example unit test.  I don't think we need it, the tests for AWS secrets and concur seem to be robust enough for our needs.

cah292 commented on src/main/java/edu/cornell/kfs/concur/web/struts/ConcurManageAccessTokenAction.java:
```diff
@@ -14,16 +18,23 @@
+    public ActionForward execute(ActionMapping mapping, ActionForm form, HttpServletRequest request,
+            HttpServletResponse response) throws Exception {
+        return getAwsSecretService().doWithAwsSecretsCachingEnabled(
+                () -> super.execute(mapping, form, request, response));
```
Yes, the intention of this override was to enable the AWS Secrets cache for all the "methodToCall"-related servlet actions on this page. I will see if I can make that intent clearer in this code.

cah292 commented on src/main/java/edu/cornell/kfs/sys/CUKFSConstants.java:
```diff
@@ -15,6 +15,7 @@
     public static final String CASH_REVERSION_OBJECT_CD = "CASH_REVERSION_OBJECT_CD";
     public static final String RULE_CODE_CA = "CA";
     public static final String DELIMITER = ".";
+    public static final String US_LOCALE_STRING = "en_US";
```
This constant is for cases when we need to reference the US locale in an annotation parameter. With parameters on Java annotations, their values have to be constant expressions (like a literal string or a concatenation of literal strings), otherwise you will get a compiler error. Thus, Locale.US cannot be used directly as an annotation parameter value, even if written to a final variable first.

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:
```diff
@@ -24,26 +28,62 @@
+                    getAwsSecretsCacheStack().pollFirst();
+                }
+            } catch (RuntimeException e) {
+                LOG.error("doWithAwsSecretsCachingEnabled: Unexpected error when clearing and removing cache", e);
```
I had concerns about the Callable's thrown exception (if any) getting lost if this nested try-catch had its error rethrown, but I agree that it would be good to rethrow this nested block's exception if it did end up with an unexpected problem. I'll rework this to add better error handling and error logging, so that we can keep better track of any exceptions thrown by the Callable and the nested cleanup block.

cah292 commented on src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java:
```diff
@@ -141,28 +195,44 @@ public void updateSecretNumber(String awsKeyName, boolean useKfsInstanceNamespac
+    protected String convertDateToString(Date date) {
         if (date == null) {
-            return StringUtils.EMPTY;
+            throw new IllegalArgumentException("Date value to be stored in AWS Secrets cannot be null");
```
I was doing some AWS Secrets testing locally, and ran into an unexpected exception from the AWSSecretsManager class when trying to store an empty string in AWS. (The AWSSecretsManager didn't like zero-length content for some reason.) A single-space string worked fine, though, and so did null values (though the nulls seem to have been treated as no-ops when saving to AWS).

If you think we should allow empty strings and null dates in AWS, then we may need some other representation of that. Perhaps we could use a special indicator value such as "NULL" or an empty JSON object ("{}"). Please let me know what your thoughts are on how to best represent such null/empty values in AWS.

jhulslander replied:
I think something "NULL" makes sense, thank you

cah292 commented:
I've committed another round of changes, to address the latest PR feedback and to also make some further improvements to the AWS secrets management. I believe I've addressed all the various areas of concern, but please let me know if you find anything else that needs to be changed.

This is ready for review again. Also keep in mind that the related "config" PR has to be approved and merged before this PR's unit tests will pass.

cah292 commented:
I also had a DevOps-related AWS secrets management question. For our KFS instances in AWS, are they currently set up yet to configure and refresh AWS security tokens regularly? In my local testing for AWS secrets management, I get expired-security-token errors if my last login for AWS CLI was not recent enough.

dp65 commented:
> I also had a DevOps-related AWS secrets management question. For our KFS instances in AWS, are they currently set up yet to configure and refresh AWS security tokens regularly? In my local testing for AWS secrets management, I get expired-security-token errors if my last login for AWS CLI was not recent enough.

@cah292 
https://docs.aws.amazon.com/secretsmanager/latest/userguide/auth-and-access_iam-policies.html

For local testing, you are using your NetID/Shibboleth credentials, covered under the "Federated users" paragraph.  The token so acquired lasts 1 hour.  
For our KFS instances in AWS, a different mechanism is used, covered under the "Applications running on EC2 paragraph".

cah292 commented:
@dp65 Thanks for the identity-policy feedback; that's very helpful. It sounds like we should be fine with the AWS Secrets Manager usage on our KFS instances.

cah292 commented:
As noted in today's bookclub meeting, we have decided to remove the AWS caching for now, and to implement a parameter-based configuration for switching between AWS Secrets Manager and the Web Service Credentials table to manage the Concur tokens. I will work on implementing those changes, and further review of this PR can wait until after I have implemented them.

cah292 commented:
Change of plans: I'll be stopping this PR implementation shortly in order to focus on the KEW-to-KFS work. I'll add some notes on the related user story about the remaining changes needed, so that another developer can bring this coding to completion.

jhulslander commented:
Close PR as it needs to be restructured per team discussion
